### ISSUE TITLE
Add homogen to R-Instat

### ISSUE THREAD

**Issue by rdstern** [africanmathsinitiative/R-Instat#8630]
@jkmusyoka we already have a place for Homogenisation in R-Instat:

<img width="294" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/8fa3e23a-1e22-42d4-a9bf-b73d098e62ea">

I suggest a further step that we use the existing Homogenization dialog and change the buttons at the top, so there becomes a new second button called Homogen - and we just have the 2 buttons.

<img width="316" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/b8c38974-15c0-42a3-83d1-93103f9a384a">

The next step will be to sort out the dialog.  It will probably need 2 data selectors, with one for the daily data and the second for the station details?

**Comment by jkmusyoka** [comment]
Yes @rdstern I checked how the function works and I agree about the two receivers, one for daily data and another for station file in the dialog

But we would first need the csv2climatol function to prepare the input file for the homogen package. Using the Western Kenya dataset as an example, the R code looks likes this:-

csv2climatol('wkenya_rain_2012.csv', datacol=1:5, stnfile='wkenya_stations_info.csv', 
             stncol=c(8,7,9,1,2),varcli='wkenya', header=TRUE)

The output would be 'wkenya' file that would be the main input into the homogen code as follows:-

homogen('wkenya', 1950, 2012)

@Vitalis95 would you like to start working on this dialogue?

**Comment by rdstern** [comment]
@Vitalis95 I hope that the csv2climatol function above might be done behind the scenes, but it way be that you want to run climatol in 2 stages?  If so, they may be from 2 radio buttons at the top of the homogenization dialog.

The Kenya data seems to be a good test case, at least for the rainfall.  I assume the starting point will be the data file as a climatic data frame.  The station date will be a second data frame, presumably with the station name or ID as a linking variable.  The stations used will just be those in the data (which may be filtered).  There will sometimes be more stations in that file that are not in the data file.

We don't have many existing dialogs with 2 selectors, but the Climatic > Check Data > Check Station Locations is one.

<img width="311" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/58e6a483-d3e6-4363-a5ba-55b042ba0e5c">

I suggest, for you, the Data File might be first and the Station File could be below?

Then discuss with @jkmusyoka what the receivers should be, and what other controls you need on the dialog.

**Comment by Vitalis95** [comment]
@jkmusyoka @rdstern , here is the sketch of the dialog. The Daily data  and Station data receivers takes column indexes instead of column names. The` check box header` is checked if the  `csvfile` has header otherwise unchecked




![20231114_113116](https://github.com/africanmathsinitiative/R-Instat/assets/89011415/4dad0db2-8c4c-4d1b-9d23-4124ccb7aaae)

**Comment by Vitalis95** [comment]
@rdstern , @jkmusyoka , The  `homogen`, function uses column indexes instead of column names,  also we had the same issue with `factominer package`
@Patowhiz , is it  possible to have an option of  choosing columns indexes ?, for now R-Instat uses columns names and 
going forward we may experience other `packages` that uses column indexes

**Comment by jkmusyoka** [comment]
Thanks @Vitalis95 for getting the dialogue started.

I would like to add the data and R code which may help further development of the dialogue. 
[wkenya_stations.csv](https://github.com/africanmathsinitiative/R-Instat/files/13388499/wkenya_stations.csv)
[wkenya_rain_2012.csv](https://github.com/africanmathsinitiative/R-Instat/files/13388479/wkenya_rain_2012.csv)
The R code is commented to help understand the steps.
[Rscript_homogen_wkenya.pdf](https://github.com/africanmathsinitiative/R-Instat/files/13388545/Rscript_homogen_wkenya.pdf)



My suggestions for the dialogue are as follows:-

1. Would it better to rename the "Homogen" tab to "Multiple"? We need a simple name that users would relate with. Some of our users may not know about the Homogen function?
2. Most daily climate files would be imported with headers so could we get rid of the checkbox and make header=T as default?
3. There are three steps before the final tests are performed on the data. Would the R commands for each of these steps be executed upon clicking the "OK" button? 
4. There is also the question of how would the output files be displayed? In R they include the Qc report, the homogen tests results. 

Back to you @Vitalis95

**Comment by john-bagiliko** [comment]
I attach here a cleaned data of the ones attached to this issue earlier by @jkmusyoka. 
[wkenya_52_metadata.csv](https://github.com/africanmathsinitiative/R-Instat/files/13658885/wkenya_52_metadata.csv)
[wkenya_52_stations_1960_2012.csv](https://github.com/africanmathsinitiative/R-Instat/files/13658886/wkenya_52_stations_1960_2012.csv). To arrive at this version of data, I did the following:

1. First, I checked whether the station names in the metadata are consistent with the ones in the data itself. I found some inconsistencies in some of the names. For instance, a station with name **HILL SCHOOL ELDORET** in the metadata is written as  **Hill School Eldoret** in daily rainfall observations data. Another named **OFFICE GAUGE** in the metadata is written as **Office Gauge**  in the observations data. To fix these issues, I removed all spaces in the station names and converted all station names to upper case in both the metadata and the observations data itself.  
2. I also checked if the number of stations in the metadata is equal to the number of stations in the actual data. I found that the observation data has 53 stations while the metadata has 52. The additional station in the data was with name **WORKSHOP**, so it was filtered out. 
3. The data (as well as the metadata) was sorted according to the station names (in alphabetical order) since **homogen** requires that the order of the stations in the data appear in the order they are given in the metadata.  
4. I created a date column from the year, month, and day columns in order to fill date gaps. Homogen requires that all stations have the same start date as well as end date.  Since this was not the case in the data provided, this was a necessary step in fixing that. 
5. Fill date gaps: I used the default settings in R-Instat's **Fill Date Gaps** dialog to fill the gaps (ensuring complete years); defined the data as Climatic (no duplicates were found), and checked the inventory. I found that before 1960, many of the stations did not have data so I filtered such that the earliest start date is **1960-01-01** (and the latest end date **2012-12-31**). Note that, with this, different stations still had different start and end dates. I needed all stations to have their start date as **1960-01-01**  and the end date as  **2012-12-31**, so I used the Fill Date Gaps dialog again, and, this time, selected the **Fixed Limits** option and set the dates as desired.  And now the inventory plot looked like this:  ![wkenya_inventory](https://github.com/africanmathsinitiative/R-Instat/assets/40699289/fcaabb42-37bb-43ae-828d-90af8d345ce4)  
6. Quality control of rainfall data: quality-controlling the variable of interest  (rain)  is a very important step before the homogenisation is done with homogen. I found about 1849 instances where the same rainfall value is recorded for 2 or 3 consecutive days.  There were also, 4995 observations where it rains continuously for 11 or more days.  It is important to check these and other aspects of the data for potential issues and fix them before the homogenization is done. I did not go much into this, but it is just to highlight it as an unskippable item in the procedure. 
7. Remove date column: I removed the date column from the data to just have station, year, month, day, and rain columns as this is desirable for the **homogen** function. 
8. Back to the metadata: I found some of the station IDs were missing. I created another column called **Codes** from the station names by just converting the station names to lowercase, which will be used as the station codes for **homogen**.  

I will give my feedback on the homogenisation dialog as a separate comment.

**Comment by Vitalis95** [comment]
@john-bagiliko , thanks for your comprehensive review

For **item** **1)** above, I will implement that
For **item 2)**  - I will check if the package is already added in R-Instat
For **item 3)**- having separate dialogs for each function is a bit tricky, each function depends on the `climatic variable` produced by the preceding function, we can have them in the same dialog and my be each function passed by a control, like a checkbox- the function only runs when we the checkbox is checked  
@rdstern , what is your take on this? 

For **item 5)**, @john-bagiliko , do you have any suggestions concerning this?

Thanks

**Comment by john-bagiliko** [comment]
@Vitalis95, for point 5, I was thinking about the possibility where a user is allowed to choose a directory for the output files. I understand this may be complicated to implement, so maybe it should not be a priority right now.
